2015/10-dancer-i18n-with-log-report.pod

```diff
@@ -0,0 +1,206 @@
+
+Let's add a translater to the application, but first, we'll need some
+translation tables.
+
+=head3 Extracting and creating PO translation tables
+
+These can be extracted using L<xgettext-perl>. From the base directory of your
+application:
+
+ xgettext-perl -p messages --from-code=iso-8859-1 .
+
+This will scan the current directory for appropriate perl files and create a
+new directory C<messages> with the strings to be translated as message IDs in a
+default C<.po> file. For more information on the module used to extract the
+strings, see L<Log::Report::Extract::PerlPPI>. Note that the domain specified
+in the options for L<Dancer2::Plugin::LogReport> (or L<Log::Report>) is used
+within the filename. Domains can be used to group together sets of messages
+(such as if you want to separate groups of messages for different modules).
+
+For the purposes of this exercise, a PO file should have been created
+containing the 2 message IDs ("Hello world" and "Hello {name}"). From this
+default PO file, create a specific language file in a new directory for this
+domain:
+
+ mkdir messages/my-domain/
+ cp messages/my-domain.utf-8.po messages/my-domain/de_DE.utf8.po
+
+Now edit the file, inserting the translation text into the existing lines (the
+"fuzzy" flag can be removed to indicate that the text has been translated):
+
+ #: ./lib/Advent.pm:11
+ #,
+ msgid "Hello world"
+ msgstr "Hallo Welt"
+ #: ./lib/Advent.pm:12
+ #,
+ msgid "Hello {name}"
+ msgstr "Hallo {name}"
+
+=head3 Adding a translator to your application
+
+To add a translator, add the following code to your route handler:
+
+ use Log::Report::Translator::POT;
+
+ (textdomain 'my-domain')->configure(
+   translator => Log::Report::Translator::POT->new(
+     lexicon => 'messages' # Directory of the PO files
+   )
+ );
+
+Now run your application, this time specifying another language:
+
+ LC_MESSAGES=de_DE.utf8 perl bin/app.pl
+
+When you visit the index page, the text should now have been appropriately
+translated.
+
+=head2 Translating text within templates
+
+Translating text within templates is very similar to translating strings within
+the application, except that being a template, the translation functions aren't
+quite so readily available. For the purposes of this example, I'm going to
+assume that you're using C<Template::Toolkit>.
+
+As we don't have a readily available translating function, we'll do things the
+other way around this time, and start by providing a translating function to each
+template. Add the following to your route handler:
+
+ sub translate {
+   my $msg = Log::Report::Message->fromTemplateToolkit('tt-domain', @_);
+   $msg->toString;
+ }
+ 
+ hook before_template => sub {
+   shift->{loc}  = \&translate;
+ };
+
+The above code adds a subroutine C<loc> that can be called from within each
+template. Note that we are using a different domain. This is normally best so
+as to keep the sets of messages completely separate, but it's also possible to
```

Comment:
We have many readers who are not native English speakers. I would avoid an expression like "so as to".

"This is normally best  as to keep the sets of messages..." -> "This is a good practice in order to keep...", maybe?